### ISSUE TITLE
Introduce errors package

### ISSUE THREAD

**Issue by marwan-at-work** [gomods/athens#275]
Athens is quite a big project with a lot of moving parts and dependencies. We should have a consistent way to construct and communicate errors through the codebase. 

We should have our own errors package with an `Error` struct holds relevant Athens information about a system error. 

An Error struct can hold information about what led to the creation of that error: module path, module operation, storage, user etc. 

Constructing an error should be easy as `errors.E(op, "error message", moduleInfo, severityLevel)`

Logging an error should be as easy as `log.SystemErr(err)` which will internally cast the error interface into the Error struct to probe the severityLevel and the operation to do structured logging. 

inspiration: https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html

**Comment by marpio** [comment]
Also from the same author: https://middlemost.com/failure-is-your-domain/

**Comment by michalpristas** [comment]
while reading the article I realized I've read it before, but it was a nice recap.
It would be great to have something like this with Athens but first, we need to find a common delimiter for our errors. Just like with the mentioned Upspin they developed the pkg with the problem in mind to solve.

We don't play with errors just yet so maybe our understanding of how we should approach this might be a bit off. But I would start with following Upspin error for implementation/API part and composing our errors of:
- Module
- Version
- Op (operation like cache-miss, source-pull, ...)
- Kind (just like with upspin)
- Err (cause of error)

So to sum it up, yeah let's 🔨 it. Nice and searchable errors rocks

**Comment by marwan-at-work** [comment]
I agree, we should at least have one canonical way to construct errors and then as we progress we'll see what kind of information/behavior it should have. But we should start conforming our errors before it gets outta hand :P 

I'll take a stab at it today or this week

**Comment by arschles** [comment]
Sounds like a plan! I see some packages where errors are pretty
standardized so I can go error-ify those and we can move in increments

**Comment by arschles** [comment]
I wonder if https://github.com/pkg/errors would be useful here?
